### PR TITLE
Fix 8947

### LINKED ISSUE #8947: Measure time outbound ops are sitting in queue due to active batch
With move toward batching being by default at JS turn granularity, we will change how long it takes from the time someone make a local change (DDS) to side-effects (op) to be sent to socket. If application spends a lot of time doing very expensive operations in response to user actions, this time might be long, and it will impact end-to-end op latency metrics we track (see first item below).

Related issues:
- #7406
- #8912
- #8911

Ideally, we track all these metrics on a single op, such that we can easily do math on these metrics.

@andre4i - FYI.

### PR DESCRIPTION
**Pull request by NicholasCouri**
Initial Fix for #8947  - create a property bag for the performance data we will submit in the opPerfRoundtripTime event.
+ Addressing part of the 8912 (excluding the tracking or pauses and resumes).

### CODE REVIEW

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 1]
```diff
@@ -18,21 +18,46 @@ import { assert, performance } from "@fluidframework/common-utils";
...
+    /** Time spent between DeltaManager's inbound "push" event until the DeltaManager "op" event */
+    "durationInboundToProcessing" |
+    /** Length of the DeltaManager's inbound queue at the time of the processing */
+    "lenghtInboundQueue";
+
+type IOpPerfTelemetryProperties =
+    Pick<ITelemetryProperties, IOpPerfTelemetryRequiredProperties>;
+
+/**
+ * We report various latency-related errors when waiting for op roundtrip takes longer than that amout of time.
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:35]: >We report various latency-related errors when waiting for op roundtrip takes longer than that amout of time [](http://example.com/codeflow?start=3&length=107)

Oppss wrong copy and paste.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 2]
```diff
@@ -18,21 +18,46 @@ import { assert, performance } from "@fluidframework/common-utils";
...
+    /** Measure time outbound op is sitting in queue due to active batch */
+    "durationOutboundQueue" |
+    /** Time spent between DeltaManager's inbound "push" event until the DeltaManager "op" event */
+    "durationInboundToProcessing" |
+    /** Length of the DeltaManager's inbound queue at the time of the processing */
+    "lenghtInboundQueue";
+
+type IOpPerfTelemetryProperties =
+    Pick<ITelemetryProperties, IOpPerfTelemetryRequiredProperties>;
+
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:33]: Not sure if this is the best way to track the perf properties "bag"  - the goal is make sure that new properties would need to be enforced here.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 3]
```diff
@@ -18,21 +18,46 @@ import { assert, performance } from "@fluidframework/common-utils";
...
+    "durationInboundQueue" |
+    /** Measure time outbound op is sitting in queue due to active batch */
+    "durationOutboundQueue" |
+    /** Time spent between DeltaManager's inbound "push" event until the DeltaManager "op" event */
+    "durationInboundToProcessing" |
+    /** Length of the DeltaManager's inbound queue at the time of the processing */
+    "lenghtInboundQueue";
+
+type IOpPerfTelemetryProperties =
+    Pick<ITelemetryProperties, IOpPerfTelemetryRequiredProperties>;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:32]: I saw this technique used in summarizer, but I'm not sure I like it. I'd rather go with proper definition of interface, which would be strongly typed.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 4]
```diff
@@ -18,21 +18,43 @@ import { assert, performance } from "@fluidframework/common-utils";
...
 
+interface IOpPerfTelemetryProperties {
+    /** Track how long op is sitting in inbound queue until it is processed */
+    durationInboundQueue: number | undefined;
+    /** Measure time outbound op is sitting in queue due to active batch */
+    durationOutboundQueue: number | undefined;
+    /** Time spent between DeltaManager's inbound "push" event until the DeltaManager "op" event */
+    durationInboundToProcessing: number | undefined;
+    /** Length of the DeltaManager's inbound queue at the time of the processing */
+    lenghtInboundQueue: number | undefined;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:29]: Worth clarifying when we collect that metric (in what phase) - is it when op is added to inbound queue, or some other time?

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 5]
```diff
@@ -18,21 +18,43 @@ import { assert, performance } from "@fluidframework/common-utils";
...
+    // To track round trip time for every %500 client message.
     private clientSequenceNumberForLatencyStatistics: number | undefined;
 
-    private opTimeSittingInInboundQueue: number | undefined;
-    private durationSittingInInboundQueue: number | undefined;
+    private readonly opProcessingTimes: number | undefined[] = [];
+
+    // Performance Data to be reported for ops round trips and processing.
+    private opPerfData: IOpPerfTelemetryProperties = {
+        durationInboundQueue: undefined,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:53]: All the members are undefined by default, so you can remove them and just have = {}

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 6]
```diff
@@ -18,21 +18,43 @@ import { assert, performance } from "@fluidframework/common-utils";
...
-
-    // To track round trip time for every %1000 client message.
-    private opSendTimeForLatencyStatistics: number | undefined;
-
+    // To track round trip time for every %500 client message.
     private clientSequenceNumberForLatencyStatistics: number | undefined;
 
-    private opTimeSittingInInboundQueue: number | undefined;
-    private durationSittingInInboundQueue: number | undefined;
+    private readonly opProcessingTimes: number | undefined[] = [];
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:49]: I'm not sure why you need an array & OpTimings. I think an object (i.e. with property names and values) is more natural way to do it and type safe.

Comment (vladsud) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:49]: Also, these are not times (aka duration), these are timestamps for start of events. I'd call it opStartTimes.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 7]
```diff
@@ -69,8 +91,14 @@ class OpPerfTelemetry {
         this.deltaManager.on("disconnect", () => {
             this.sequenceNumberForMsnTracking = undefined;
             this.clientSequenceNumberForLatencyStatistics = undefined;
-            this.opTimeSittingInInboundQueue = undefined;
-            this.durationSittingInInboundQueue = undefined;
+            this.opProcessingTimes[OpTimings.opTimeSittingInboundQueue] = undefined;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:94]: It's better to allocate fully new object, i.e. this.opProcessingTimes = [], or this.opProcessingTimes.length = 0 (assuming it's Ok to clear all thing).

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 4 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:29]: Added the reference to the Stage.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 8]
```diff
@@ -18,21 +18,43 @@ import { assert, performance } from "@fluidframework/common-utils";
  */
 export const latencyThreshold = 5000;
 
+interface IOpPerfTelemetryProperties {
+    /** Track how long op is sitting in inbound queue until it is processed */
+    durationInboundQueue: number | undefined;
+    /** Measure time outbound op is sitting in queue due to active batch */
+    durationOutboundQueue: number | undefined;
+    /** Time spent between DeltaManager's inbound "push" event until the DeltaManager "op" event */
+    durationInboundToProcessing: number | undefined;
```
[the remark this answers is not in the document]
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:27]: Added the 5 stages from the bug

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 6 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:49]: changed to startTimes.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 7 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:94]: I believe it is.

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 6 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:49]: As for the Object, Is it ok to create as Partial<Interface> ?

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 9]
```diff
@@ -18,24 +18,34 @@ import { assert, performance } from "@fluidframework/common-utils";
...
 export const latencyThreshold = 5000;
 
+// Stages in OpPerfTelemetry that fire in the following oder:
+// 1) "submitOp" event
+// 2) DeltaManager's outbound "push" event
+// 3) DeltaManager's inbound "push" event
+// 4) Start of op processing (no explicit event for)
+// 5) DeltaManager "op" event(time to process an op)
 interface IOpPerfTelemetryProperties {
-    /** Track how long op is sitting in inbound queue until it is processed */
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts]: It is fine as it is (i.e. feel free to leave it as you have it), but I'll share how I think :)
I'd probably describe stages as long processes. I.e.
Stages:
1. Op sits in buffer in DeltaManager queue, then in outbound queue for some time
   - We do not differentiate these two today in telemetry, but first one is due to batches, second one might happen due to outbound queue being paused
3. Op is sent to service and back
4. op sits in inbound queue
5. Op is processed.

I.e. events follow (not the primary way to describe processes), and events fire in between phases (i.e. they clearly denote the boundary between phases).

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 10]
```diff
@@ -18,21 +18,48 @@ import { assert, performance } from "@fluidframework/common-utils";
  */
 export const latencyThreshold = 5000;
 
+// Stages in OpPerfTelemetry that fire in the following oder:
+// 1) "submitOp" event
+// 2) DeltaManager's outbound "push" event
+// 3) DeltaManager's inbound "push" event
+// 4) Start of op processing (no explicit event for)
+// 5) DeltaManager "op" event(time to process an op)
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:26]: It is fine as it is (i.e. feel free to leave it as you have it), but I'll share how I think :)
I'd probably describe stages as long processes. I.e.
Stages:

1. Op sits in buffer in DeltaManager queue, then in outbound queue for some time
   - We do not differentiate these two today in telemetry, but first one is due to batches, second one might happen due to outbound queue being paused
2. Op is sent to service and back
3. op sits in inbound queue
4. Op is processed.

I.e. events follow (not the primary way to describe processes), and events fire in between phases (i.e. they clearly denote the boundary between phases).

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 11]
```diff
@@ -18,21 +18,48 @@ import { assert, performance } from "@fluidframework/common-utils";
...
-
-    // To track round trip time for every %1000 client message.
-    private opSendTimeForLatencyStatistics: number | undefined;
-
+    // To track round trip time for every %500 client message.
     private clientSequenceNumberForLatencyStatistics: number | undefined;
 
-    private opTimeSittingInInboundQueue: number | undefined;
-    private durationSittingInInboundQueue: number | undefined;
+    private opProcessingTimes: Partial<IOpPerfTimings> = {};
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:59]: I think Partial can be dropped here, as all properties have default value of undefined - compiler should be happen.
Or you can use Partial, but then you can drop undefined in IOpPerfTimings definition.

Comment (vladsud) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:59]: Same for opPerfData

File: packages/runtime/container-runtime/src/connectionTelemetry.ts [hunk 12]
```diff
@@ -201,10 +249,10 @@ class OpPerfTelemetry {
                 category,
                 pingLatency: this.pingLatency,
                 msnDistance: this.deltaManager.lastSequenceNumber - this.deltaManager.minimumSequenceNumber,
-                durationInboundQueue: this.durationSittingInInboundQueue,
+                ...this.opPerfData,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/connectionTelemetry.ts:252]: Can you please check in debugger (using samples) that
- all numbers make sense (maybe inject delays here and there to validate we are recording them properly)
- that numbers are integers (if not, we might way to use TelemetryLogger.formatTick to make it show up nicer in telemetry).
-

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/connectionTelemetry.ts (hunk 12 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/connectionTelemetry.ts:252]: Yes, I've been inspecting them in the debugger and they look nice (integers).

**Pull request merged**